vendor/noise_handshake/src/handshake.rs

```diff
@@ -0,0 +1,217 @@
+                    },
+                ..
+            } => {
+                while written_bytes < len {
+                    *written_bytes = try_ready!(io.poll_write(&buffer[*written_bytes..*len]));
+                }
+                if noise.is_handshake_finished() {
+                    Ok(Async::Ready(self.wrap_up()))
+                } else {
+                    self.next = Step::read();
+                    self.poll()
+                }
+            }
+            Self {
+                noise: Some(ref mut noise),
+                io: Some(ref mut io),
+                next:
+                    Step::Read {
+                        mut enc_buffer,
+                        ref mut len,
+                    },
+            } => {
+                let mut dec_buffer = [0u8; NOISE_MAX_SIZE];
+
+                *len += try_ready!(io.poll_read(&mut enc_buffer[*len..]));
+                if *len > 0 {
+                    match noise.read_message(&enc_buffer[..*len], &mut dec_buffer) {
+                        Ok(_) => self.next = Step::write(),
+                        Err(e) => debug!("Error decoding message: {:?}", e),
+                    };
+                }
+                if noise.is_handshake_finished() {
+                    Ok(Async::Ready(self.wrap_up()))
+                } else {
+                    self.poll()
+                }
+            }
+            Self { noise: None, .. } | Self { io: None, .. } => {
+                panic!("Future is already resolved!");
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+
+    fn setup() -> (
+        NoiseHandshake<memsocket::UnboundedSocket>,
+        NoiseHandshake<memsocket::UnboundedSocket>,
+    ) {
+        let builder_resp = snow::Builder::new("Noise_XK_25519_ChaChaPoly_BLAKE2s".parse().unwrap());
+        let static_keypair_resp = builder_resp.generate_keypair().unwrap();
+        let noise_resp = builder_resp
+            .local_private_key(&static_keypair_resp.private)
+            .build_responder()
+            .unwrap();
+
+        let builder_init = snow::Builder::new("Noise_XK_25519_ChaChaPoly_BLAKE2s".parse().unwrap());
+        let static_key_init = builder_init.generate_keypair().unwrap().private;
+        let noise_init = builder_init
+            .local_private_key(&static_key_init)
+            .remote_public_key(&static_keypair_resp.public) // The initiator already knows the responder public key
+            .build_initiator()
+            .unwrap();
+
+        let (socket_init, socket_resp) = memsocket::unbounded();
+
+        let handshake_init = noise_init.init_hand_shake(socket_init);
+
+        let handshake_resp = noise_resp.init_hand_shake(socket_resp);
+
+        (handshake_init, handshake_resp)
+    }
+
+    #[test]
+    fn handshake() -> Result<(), std::io::Error> {
+        let (hs_init, hs_resp) = setup();
+
+        let mut runtime = tokio::runtime::Builder::new()
+            .core_threads(2)
```

Comment:
❓ Ahhh is this how you go it to pass travis? When I suggested this I was thinking you would always set it to 1 and then find a solution. But I guess this works 😅